File: backend/app/core/token_utils.py
"""
Token Counting Utilities

Provides consistent token counting across the application using character-based approximation.
1 token ≈ 4 characters for simplicity and performance.
"""

from typing import List, Dict, Any, Optional, Union
import re

class TokenCounter:
    """
    Token counting utility using character-based approximation.
    Provides consistent token counting across all application components.
    """
    
    # Token approximation: 1 token ≈ 4 characters
    CHARS_PER_TOKEN = 4
    
    @classmethod
    def count_tokens(cls, text: str) -> int:
        """
        Count tokens in text using character-based approximation.
        
        Args:
            text: Text to count tokens for
            
        Returns:
            Estimated token count
        """
        if not text:
            return 0
        
        # Clean up text (remove excessive whitespace, normalize)
        cleaned_text = re.sub(r'\s+', ' ', text.strip())
        char_count = len(cleaned_text)
        
        # Convert to tokens using approximation
        token_count = max(1, (char_count + cls.CHARS_PER_TOKEN - 1) // cls.CHARS_PER_TOKEN)
        
        return token_count
    
    @classmethod
    def count_tokens_in_messages(cls, messages: List[Dict[str, Any]]) -> int:
        """
        Count total tokens in a list of messages.
        
        Args:
            messages: List of message dictionaries with 'content' field
            
        Returns:
            Total token count across all messages
        """
        total_tokens = 0
        
        for message in messages:
            content = message.get('content', '')
            if isinstance(content, str):
                total_tokens += cls.count_tokens(content)
        
        return total_tokens
    
    @classmethod
    def count_tokens_in_documents(cls, documents: List[Dict[str, Any]]) -> int:
        """
        Count total tokens in a list of documents.
        
        Args:
            documents: List of document dictionaries
            
        Returns:
            Total token count across all documents
        """
        total_tokens = 0
        
        for doc in documents:
            # Try different content fields
            content = (
                doc.get('extracted_text') or 
                doc.get('content') or 
                doc.get('summary') or 
                ''
            )
            
            if isinstance(content, str):
                total_tokens += cls.count_tokens(content)
        
        return total_tokens
    
    @classmethod
    def count_tokens_in_notes(cls, notes: List[Dict[str, Any]]) -> int:
        """
        Count total tokens in a list of paper notes.
        
        Args:
            notes: List of paper note dictionaries
            
        Returns:
            Total token count across all notes
        """
        total_tokens = 0
        
        for note in notes:
            content = note.get('content', '') or note.get('text', '')
            if isinstance(content, str):
                total_tokens += cls.count_tokens(content)
        
        return total_tokens
    
    @classmethod
    def truncate_to_token_limit(cls, text: str, max_tokens: int) -> str:
        """
        Truncate text to stay within token limit.
        
        Args:
            text: Text to truncate
            max_tokens: Maximum tokens allowed
            
        Returns:
            Truncated text that fits within token limit
        """
        if not text:
            return text
        
        current_tokens = cls.count_tokens(text)
        
        if current_tokens <= max_tokens:
            return text
        
        # Calculate target character count
        max_chars = max_tokens * cls.CHARS_PER_TOKEN
        
        # Truncate to approximate character limit
        if len(text) > max_chars:
            truncated = text[:max_chars].rsplit(' ', 1)[0]  # Break at word boundary
            return truncated + "..."
        
        return text
    
    @classmethod
    def filter_messages_by_token_limit(
        cls, 
        messages: List[Dict[str, Any]], 
        max_tokens: int,
        reverse_chronological: bool = True
    ) -> List[Dict[str, Any]]:
        """
        Filter messages to stay within token limit, keeping most recent first.
        
        Args:
            messages: List of message dictionaries
            max_tokens: Maximum total tokens allowed
            reverse_chronological: If True, keep newest messages first
            
        Returns:
            Filtered list of messages within token limit
        """
        if not messages:
            return []
        
        # Sort messages by preference (newest first by default)
        sorted_messages = messages.copy()
        if reverse_chronological:
            sorted_messages.reverse()
        
        filtered_messages = []
        current_tokens = 0
        
        for message in sorted_messages:
            message_tokens = cls.count_tokens(message.get('content', ''))
            
            if current_tokens + message_tokens <= max_tokens:
                filtered_messages.append(message)
                current_tokens += message_tokens
            else:
                # Stop adding messages once we hit the limit
                break
        
        # Restore original order if we reversed it
        if reverse_chronological:
            filtered_messages.reverse()
        
        return filtered_messages
    
    @classmethod
    def filter_items_by_token_limit(
        cls,
        items: List[Dict[str, Any]],
        max_tokens: int,
        content_field: str = 'content',
        keep_pinned: bool = False,
        pinned_field: str = 'is_pinned'
    ) -> tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
        """
        Filter items by token limit, optionally keeping pinned items.
        
        Args:
            items: List of item dictionaries  
            max_tokens: Maximum total tokens allowed
            content_field: Field name containing the text content
            keep_pinned: Whether to always include pinned items
            pinned_field: Field name indicating if item is pinned
            
        Returns:
            Tuple of (included_items, excluded_items)
        """
        if not items:
            return [], []
        
        included_items = []
        excluded_items = []
        current_tokens = 0
        
        # First pass: Always include all pinned items and count their tokens
        pinned_tokens = 0
        for item in items:
            is_pinned = item.get(pinned_field, False)
            if is_pinned and keep_pinned:
                item_tokens = cls.count_tokens(item.get(content_field, ''))
                included_items.append(item)
                pinned_tokens += item_tokens
        
        # Second pass: Add unpinned items until total (pinned + unpinned) exceeds limit
        current_tokens = pinned_tokens
        for item in items:
            is_pinned = item.get(pinned_field, False)
            if not is_pinned:  # Only process unpinned items
                item_tokens = cls.count_tokens(item.get(content_field, ''))
                
                if current_tokens + item_tokens <= max_tokens:
                    # Include unpinned items that fit within total limit
                    included_items.append(item)
                    current_tokens += item_tokens
                else:
                    # Exclude unpinned items that exceed total limit
                    excluded_items.append(item)
        
        return included_items, excluded_items
    
    @classmethod
    def is_over_token_limit(cls, text: str, max_tokens: int) -> bool:
        """
        Check if text exceeds token limit.
        
        Args:
            text: Text to check
            max_tokens: Maximum tokens allowed
            
        Returns:
            True if text exceeds limit
        """
        return cls.count_tokens(text) > max_tokens
    
    @classmethod
    def get_token_usage_info(cls, text: str, max_tokens: int) -> Dict[str, Any]:
        """
        Get detailed token usage information.
        
        Args:
            text: Text to analyze
            max_tokens: Maximum tokens allowed
            
        Returns:
            Dictionary with usage information
        """
        current_tokens = cls.count_tokens(text)
        
        return {
            'current_tokens': current_tokens,
            'max_tokens': max_tokens,
            'remaining_tokens': max(0, max_tokens - current_tokens),
            'over_limit': current_tokens > max_tokens,
            'usage_percentage': min(100.0, (current_tokens / max_tokens) * 100.0) if max_tokens > 0 else 0.0
        }
    
    @classmethod
    def get_conversation_token_analysis(
        cls, 
        messages: List[Dict[str, Any]], 
        max_tokens: int,
        threshold_percentage: float = 0.8
    ) -> Dict[str, Any]:
        """
        Analyze conversation token usage and determine if compacting is needed.
        
        Args:
            messages: List of conversation messages
            max_tokens: Maximum tokens allowed for conversation
            threshold_percentage: Threshold percentage (0.8 = 80%) to trigger compacting
            
        Returns:
            Dictionary with analysis results
        """
        if not messages:
            return {
                'total_tokens': 0,
                'max_tokens': max_tokens,
                'usage_percentage': 0.0,
                'needs_compacting': False,
                'threshold_tokens': int(max_tokens * threshold_percentage),
                'over_threshold': False,
                'message_count': 0,
                'per_message_tokens': []
            }
        
        # Calculate tokens per message
        per_message_tokens = []
        total_tokens = 0
        
        for i, message in enumerate(messages):
            content = message.get('content', '')
            message_tokens = cls.count_tokens(content)
            per_message_tokens.append({
                'index': i,
                'sender': message.get('sender', 'unknown'),
                'tokens': message_tokens,
                'content_preview': content[:100] + ('...' if len(content) > 100 else '')
            })
            total_tokens += message_tokens
        
        threshold_tokens = int(max_tokens * threshold_percentage)
        usage_percentage = (total_tokens / max_tokens) * 100.0 if max_tokens > 0 else 0.0
        over_threshold = total_tokens > threshold_tokens
        
        return {
            'total_tokens': total_tokens,
            'max_tokens': max_tokens,
            'usage_percentage': usage_percentage,
            'needs_compacting': over_threshold,
            'threshold_tokens': threshold_tokens,
            'over_threshold': over_threshold,
            'message_count': len(messages),
            'per_message_tokens': per_message_tokens
        }
    
    @classmethod
    def select_messages_for_compacting(
        cls,
        messages: List[Dict[str, Any]],
        messages_to_keep_recent: int = 2,
        existing_summary: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Select which messages to compact and which to keep recent.
        
        Args:
            messages: List of conversation messages
            messages_to_keep_recent: Number of recent messages to keep uncompacted
            existing_summary: Existing summary from previous compacting
            
        Returns:
            Dictionary with messages to compact and messages to keep
        """
        if len(messages) <= messages_to_keep_recent:
            return {
                'messages_to_compact': [],
                'messages_to_keep': messages,
                'existing_summary': existing_summary,
                'can_compact': False
            }
        
        # Split messages: older ones for compacting, recent ones to keep
        messages_to_compact = messages[:-messages_to_keep_recent] if messages_to_keep_recent > 0 else messages
        messages_to_keep = messages[-messages_to_keep_recent:] if messages_to_keep_recent > 0 else []
        
        return {
            'messages_to_compact': messages_to_compact,
            'messages_to_keep': messages_to_keep,
            'existing_summary': existing_summary,
            'can_compact': len(messages_to_compact) > 0,
            'compacting_count': len(messages_to_compact),
            'keeping_count': len(messages_to_keep)
        }


# Create singleton instance for easy importing
token_counter = TokenCounter()

# Export commonly used functions for direct access
count_tokens = TokenCounter.count_tokens
count_tokens_in_messages = TokenCounter.count_tokens_in_messages
count_tokens_in_documents = TokenCounter.count_tokens_in_documents
truncate_to_token_limit = TokenCounter.truncate_to_token_limit
is_over_token_limit = TokenCounter.is_over_token_limit

__all__ = [
    'TokenCounter',
    'token_counter',
    'count_tokens',
    'count_tokens_in_messages', 
    'count_tokens_in_documents',
    'truncate_to_token_limit',
    'is_over_token_limit',
]

File: src/core/utils/tokenUtils.ts
/**
 * Token counting and validation utilities for memory management system
 */

// Simple token estimation based on character count
// This is an approximation - 1 token ≈ 4 characters for most text
const CHARS_PER_TOKEN = 4;

export function estimateTokenCount(text: string): number {
  if (!text) return 0;
  
  // Remove extra whitespace and normalize
  const normalizedText = text.trim().replace(/\s+/g, ' ');
  
  // Estimate tokens (rough approximation)
  return Math.ceil(normalizedText.length / CHARS_PER_TOKEN);
}

export function validateTokenLimit(text: string, maxTokens: number): {
  isValid: boolean;
  currentTokens: number;
  maxTokens: number;
  remainingTokens: number;
} {
  const currentTokens = estimateTokenCount(text);
  const remainingTokens = Math.max(0, maxTokens - currentTokens);
  
  return {
    isValid: currentTokens <= maxTokens,
    currentTokens,
    maxTokens,
    remainingTokens
  };
}

export function truncateToTokenLimit(text: string, maxTokens: number): string {
  const validation = validateTokenLimit(text, maxTokens);
  
  if (validation.isValid) {
    return text;
  }
  
  // Truncate text to fit within token limit
  const maxChars = maxTokens * CHARS_PER_TOKEN;
  const truncated = text.substring(0, maxChars);
  
  // Try to break at a word boundary to avoid cutting words in half
  const lastSpaceIndex = truncated.lastIndexOf(' ');
  if (lastSpaceIndex > maxChars * 0.8) {
    return truncated.substring(0, lastSpaceIndex);
  }
  
  return truncated;
}

export function formatTokenCount(tokens: number): string {
  if (tokens >= 1000) {
    return `${(tokens / 1000).toFixed(1)}k`;
  }
  return tokens.toString();
}

// Memory management constants - Updated to match new token limits
export const MEMORY_LIMITS = {
  // New token-based limits - UPDATED to match backend context_limits.py
  AGENT_DESCRIPTION: 4000,           // 4k tokens for agent description
  MEDIUM_TERM_MEMORY: 500,          // 500 tokens for medium term memory (temporarily for testing)
  PAPER_NOTES: 4000,                 // 4k tokens for paper notes context
  CONVERSATION_CONTEXT: 4000,        // 4k tokens for conversation messages
  
  // Document limits - UPDATED to match backend
  DOCUMENT_PER_DOCUMENT: 20000,      // 20k tokens per individual document
  AGENT_DOCUMENTS_MAX_COUNT: 5,      // Max 5 documents per agent (FIXED from 10)
  CONVERSATION_DOCUMENTS_CUMULATIVE: 20000, // 20k tokens cumulative for conversation docs
  
  // Voice system instruction - UPDATED to match backend
  VOICE_SYSTEM_INSTRUCTION: 16000,   // 16k tokens for voice system instruction
  
  // Document layer limits - ADDED from backend
  DOCUMENT_LAYER1_SUMMARY: 600,      // Layer 1 summary max tokens
  DOCUMENT_LAYER1_WORD_BANK: 200,    // Layer 1 word bank max tokens  
  DOCUMENT_LAYER1_TOTAL: 800,        // Total Layer 1 per document
  DOCUMENT_LAYER2_SUMMARY: 2000,     // Layer 2 comprehensive summary
  
  // Legacy limits (kept for backward compatibility)
  DOCUMENT_SUMMARY: 700,
  DOCUMENT_WORD_BANK: 300,
  DOCUMENT_FACTS: 2000,
  DOCUMENT_FULL: 50000,
  TOTAL_SYSTEM_MAX: 200000
} as const;

// Interfaces for new token validation
export interface TokenUsageInfo {
  currentTokens: number;
  maxTokens: number;
  remainingTokens: number;
  overLimit: boolean;
  usagePercentage: number;
}

export interface Message {
  content?: string;
  [key: string]: any;
}

export interface Document {
  extracted_text?: string;
  content?: string;
  summary?: string;
  [key: string]: any;
}

export interface Note {
  content?: string;
  text?: string;
  is_pinned?: boolean;
  [key: string]: any;
}

/**
 * Enhanced token utilities for the new token limits system
 */
export class TokenCounter {
  /**
   * Count total tokens in a list of messages.
   */
  static countTokensInMessages(messages: Message[]): number {
    let totalTokens = 0;
    for (const message of messages) {
      const content = message.content || '';
      if (typeof content === 'string') {
        totalTokens += estimateTokenCount(content);
      }
    }
    return totalTokens;
  }

  /**
   * Count total tokens in a list of documents.
   */
  static countTokensInDocuments(documents: Document[]): number {
    let totalTokens = 0;
    for (const doc of documents) {
      const content = doc.extracted_text || doc.content || doc.summary || '';
      if (typeof content === 'string') {
        totalTokens += estimateTokenCount(content);
      }
    }
    return totalTokens;
  }

  /**
   * Count total tokens in a list of notes.
   */
  static countTokensInNotes(notes: Note[]): number {
    let totalTokens = 0;
    for (const note of notes) {
      const content = note.content || note.text || '';
      if (typeof content === 'string') {
        totalTokens += estimateTokenCount(content);
      }
    }
    return totalTokens;
  }

  /**
   * Filter messages to stay within token limit, keeping most recent first.
   */
  static filterMessagesByTokenLimit(
    messages: Message[],
    maxTokens: number,
    reverseChronological: boolean = true
  ): Message[] {
    if (!messages.length) return [];

    const sortedMessages = [...messages];
    if (reverseChronological) {
      sortedMessages.reverse();
    }

    const filteredMessages: Message[] = [];
    let currentTokens = 0;

    for (const message of sortedMessages) {
      const messageTokens = estimateTokenCount(message.content || '');
      if (currentTokens + messageTokens <= maxTokens) {
        filteredMessages.push(message);
        currentTokens += messageTokens;
      } else {
        break; // Stop adding messages once we hit the limit
      }
    }

    if (reverseChronological) {
      filteredMessages.reverse();
    }

    return filteredMessages;
  }

  /**
   * Filter items by token limit, optionally keeping pinned items.
   */
  static filterItemsByTokenLimit<T extends Record<string, any>>(
    items: T[],
    maxTokens: number,
    contentField: string = 'content',
    keepPinned: boolean = false,
    pinnedField: string = 'is_pinned'
  ): { includedItems: T[]; excludedItems: T[] } {
    if (!items.length) {
      return { includedItems: [], excludedItems: [] };
    }

    const includedItems: T[] = [];
    const excludedItems: T[] = [];
    let currentTokens = 0;

    // First pass: Add all pinned items if keepPinned is true
    if (keepPinned) {
      for (const item of items) {
        if (item[pinnedField]) {
          const itemTokens = estimateTokenCount(item[contentField] || '');
          includedItems.push(item);
          currentTokens += itemTokens;
        }
      }
    }

    // Second pass: Add non-pinned items until we hit the limit
    for (const item of items) {
      if (keepPinned && item[pinnedField]) {
        continue; // Skip already included pinned items
      }

      const itemTokens = estimateTokenCount(item[contentField] || '');

      if (currentTokens + itemTokens <= maxTokens) {
        includedItems.push(item);
        currentTokens += itemTokens;
      } else {
        excludedItems.push(item);
      }
    }

    return { includedItems, excludedItems };
  }

  /**
   * Get detailed token usage information.
   */
  static getTokenUsageInfo(text: string, maxTokens: number): TokenUsageInfo {
    const currentTokens = estimateTokenCount(text);
    return {
      currentTokens,
      maxTokens,
      remainingTokens: Math.max(0, maxTokens - currentTokens),
      overLimit: currentTokens > maxTokens,
      usagePercentage: maxTokens > 0 ? Math.min(100, (currentTokens / maxTokens) * 100) : 0
    };
  }

  /**
   * Validate if a document would exceed size limits.
   */
  static validateDocumentTokens(
    content: string,
    maxTokensPerDocument: number
  ): { valid: boolean; error?: string; tokens: number } {
    const tokens = estimateTokenCount(content);
    
    if (tokens > maxTokensPerDocument) {
      return {
        valid: false,
        error: `Document exceeds ${maxTokensPerDocument.toLocaleString()} token limit`,
        tokens
      };
    }
    
    return { valid: true, tokens };
  }

  /**
   * Validate if adding documents would exceed cumulative limits.
   */
  static validateCumulativeDocumentTokens(
    existingDocuments: Document[],
    newDocuments: Document[],
    maxCumulativeTokens: number
  ): { valid: boolean; error?: string; currentTokens: number; newTokens: number } {
    const currentTokens = this.countTokensInDocuments(existingDocuments);
    const newTokens = this.countTokensInDocuments(newDocuments);
    const totalTokens = currentTokens + newTokens;
    
    if (totalTokens > maxCumulativeTokens) {
      return {
        valid: false,
        error: `Total documents would exceed ${maxCumulativeTokens.toLocaleString()} token limit`,
        currentTokens,
        newTokens
      };
    }
    
    return { valid: true, currentTokens, newTokens };
  }

  /**
   * Validate agent document limits (count and individual size).
   */
  static validateAgentDocuments(
    existingDocuments: Document[],
    newDocuments: Document[],
    maxDocuments: number,
    maxTokensPerDocument: number
  ): { valid: boolean; error?: string } {
    // Check document count limit
    if (existingDocuments.length + newDocuments.length > maxDocuments) {
      return {
        valid: false,
        error: `Maximum ${maxDocuments} documents per agent`
      };
    }

    // Check individual document token limits
    for (const doc of newDocuments) {
      const content = doc.extracted_text || doc.content || doc.summary || '';
      const validation = this.validateDocumentTokens(content, maxTokensPerDocument);
      if (!validation.valid) {
        return validation;
      }
    }

    return { valid: true };
  }
}

// Export enhanced functions
export const {
  countTokensInMessages,
  countTokensInDocuments,
  countTokensInNotes,
  filterMessagesByTokenLimit,
  filterItemsByTokenLimit,
  getTokenUsageInfo,
  validateDocumentTokens,
  validateCumulativeDocumentTokens,
  validateAgentDocuments,
} = TokenCounter;